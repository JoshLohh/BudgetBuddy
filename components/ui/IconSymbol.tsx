// Fallback for using MaterialIcons on Android and web.

import MaterialIcons from '@expo/vector-icons/MaterialIcons';
import { SymbolWeight, SymbolViewProps } from 'expo-symbols';
import { ComponentProps } from 'react';
import { OpaqueColorValue, type StyleProp, type TextStyle } from 'react-native';
import { View } from 'react-native';

type IconMapping = typeof MAPPING;
//type IconMapping = Record<SymbolViewProps['name'], ComponentProps<typeof MaterialIcons>['name']>;
//type IconSymbolName = keyof typeof MAPPING;
type IconSymbolName = keyof IconMapping;

/**
 * Add your SF Symbols to Material Icons mappings here.
 * - see Material Icons in the [Icons Directory](https://icons.expo.fyi).
 * - see SF Symbols in the [SF Symbols](https://developer.apple.com/sf-symbols/) app.
 */
const MAPPING = {
  'house.fill': 'home',
  'paperplane.fill': 'send',
  'chevron.left.forwardslash.chevron.right': 'code',
  'chevron.right': 'chevron-right',
  'plus.square.fill.on.square.fill': 'square',
  'person.3.fill': 'groups',
  'person.fill': 'add',

} as const;

/**
 * An icon component that uses native SF Symbols on iOS, and Material Icons on Android and web.
 * This ensures a consistent look across platforms, and optimal resource usage.
 * Icon `name`s are based on SF Symbols and require manual mapping to Material Icons.
 */
export function IconSymbol({
  name,
  size = 24,
  color,
  style,
}: {
  name: IconSymbolName;
  size?: number;
  color: string | OpaqueColorValue;
  style?: StyleProp<TextStyle>;
  weight?: SymbolWeight;
}) {
  //return <MaterialIcons testID="material-icon" color={color} size={size} name={MAPPING[name]} style={style} />;
  return (
    <MaterialIcons
      testID="material-icon"  // Move testID here
      color={color}
      size={size}
      name={MAPPING[name]}
      style={style}
    />
  );
}
