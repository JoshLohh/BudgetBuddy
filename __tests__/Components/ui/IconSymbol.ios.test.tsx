// IconSymbol.ios.test.tsx

import React from 'react'
import { render } from '@testing-library/react-native'
import { SymbolView, SymbolViewProps } from 'expo-symbols'
import { ViewStyle } from 'react-native'
import { IconSymbol } from '../../../components/ui/IconSymbol.ios'

jest.mock('expo-symbols', () => ({
  SymbolView: jest.fn(() => null),
}))

describe('IconSymbol (iOS)', () => {
  const defaultProps = {
    name: 'star' as SymbolViewProps['name'],
    color: '#ff0',
  }

  beforeEach(() => {
    (SymbolView as jest.Mock).mockClear()
  })

  it('renders with required props and defaults', () => {
    render(<IconSymbol {...defaultProps} />)
    expect(SymbolView).toHaveBeenCalledWith(
      expect.objectContaining({
        name: 'star',
        tintColor: '#ff0',
        weight: 'regular',
        resizeMode: 'scaleAspectFit',
        style: expect.arrayContaining([
          expect.objectContaining({ height: 24, width: 24 }),
          undefined,
        ]),
      }),
      undefined // Fix here
    )
  })

  it('forwards size, color, style, weight', () => {
    const style: ViewStyle = { margin: 25 }
    render(
      <IconSymbol
        name={'bolt' as SymbolViewProps['name']}
        size={32}
        color="#123"
        style={style}
        weight="bold"
      />
    )
    expect(SymbolView).toHaveBeenCalledWith(
      expect.objectContaining({
        name: 'bolt',
        tintColor: '#123',
        weight: 'bold',
        resizeMode: 'scaleAspectFit',
        style: expect.arrayContaining([
          expect.objectContaining({ height: 32, width: 32 }),
          style,
        ]),
      }),
      undefined // Fix here
    )
  })

  it('handles absence of optional props', () => {
    render(<IconSymbol name={'cloud' as SymbolViewProps['name']} color="#00f" />)
    expect(SymbolView).toHaveBeenCalledWith(
      expect.objectContaining({
        name: 'cloud',
        tintColor: '#00f',
        weight: 'regular',
        resizeMode: 'scaleAspectFit',
        style: expect.arrayContaining([
          expect.objectContaining({ height: 24, width: 24 }),
          undefined,
        ]),
      }),
      undefined // Fix here
    )
  })

  //Snapshot
  it('matches snapshot with custom props', () => {
    const tree = render(
      <IconSymbol
        name={'wifi' as SymbolViewProps['name']}
        size={28}
        color="#aaa"
        style={{ alignSelf: 'center' }}
        weight="semibold"
      />
    ).toJSON()
    expect(tree).toMatchSnapshot()
  })
})
