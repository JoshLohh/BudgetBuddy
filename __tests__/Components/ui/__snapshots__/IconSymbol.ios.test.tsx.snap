// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IconSymbol (iOS) matches snapshot with custom props 1`] = `null`;
